br/pkg/restore/misc.go

```diff
@@ -44,6 +53,182 @@ const (
+		return 0, 0, false
+	}
+	snapshotBackupTs = ts
+	return restoreCommitTs, snapshotBackupTs, true
+}
+
+func (markerFile *LogRestoreTableIDsMarkerFile) checksumLogRestoreTableIDsMarkerFile() []byte {
+	hasher := sha256.New()
+	hasher.Write(binary.LittleEndian.AppendUint64(nil, markerFile.RestoreCommitTs))
+	hasher.Write(binary.LittleEndian.AppendUint64(nil, markerFile.SnapshotBackupTs))
+	for _, tableId := range markerFile.TableIds {
+		hasher.Write(binary.LittleEndian.AppendUint64(nil, uint64(tableId)))
+	}
+	return hasher.Sum(nil)
+}
+
+func (markerFile *LogRestoreTableIDsMarkerFile) setChecksumLogRestoreTableIDsMarkerFile() {
+	markerFile.Checksum = markerFile.checksumLogRestoreTableIDsMarkerFile()
+}
+
+// MarshalLogRestoreTableIDsMarkerFile generates an markerfile and marshals it. It returns its filename and the marshaled data.
+func MarshalLogRestoreTableIDsMarkerFile(restoreCommitTs, snapshotBackupTs uint64, tableIds []int64) (string, []byte, error) {
+	markerFile := &LogRestoreTableIDsMarkerFile{
+		RestoreCommitTs:  restoreCommitTs,
+		SnapshotBackupTs: snapshotBackupTs,
+		TableIds:         tableIds,
+	}
+	markerFile.setChecksumLogRestoreTableIDsMarkerFile()
+	filename := markerFile.filename()
+	data, err := proto.Marshal(markerFile)
+	if err != nil {
+		return "", nil, errors.Trace(err)
+	}
+	return filename, data, nil
+}
+
+// unmarshalLogRestoreTableIDsMarkerFile unmarshals the given markerfile.
+func unmarshalLogRestoreTableIDsMarkerFile(data []byte) (restoreCommitTs, snapshotBackupTs uint64, tableIds []int64, err error) {
+	markerFile := &LogRestoreTableIDsMarkerFile{}
+	if err = proto.Unmarshal(data, markerFile); err != nil {
+		return 0, 0, nil, errors.Trace(err)
+	}
+	if !bytes.Equal(markerFile.checksumLogRestoreTableIDsMarkerFile(), markerFile.Checksum) {
+		return 0, 0, nil, errors.Errorf(
+			"checksum mismatch (calculated checksum is %s but the recorded checksum is %s), the log restore table IDs blocklist file may be corrupted",
+			base64.StdEncoding.EncodeToString(markerFile.checksumLogRestoreTableIDsMarkerFile()),
+			base64.StdEncoding.EncodeToString(markerFile.Checksum),
+		)
+	}
+	return markerFile.RestoreCommitTs, markerFile.SnapshotBackupTs, markerFile.TableIds, nil
+}
+
+func fastWalkLogRestoreTableIDsMarkerFile(
+	ctx context.Context,
+	s storage.ExternalStorage,
+	filterOutFn func(restoreCommitTs, snapshotBackupTs uint64) bool,
+	executionFn func(ctx context.Context, filename string, restoreCommitTs uint64, tableIds []int64) error,
+) error {
+	filenames := make([]string, 0)
+	if err := s.WalkDir(ctx, &storage.WalkOption{SubDir: logRestoreTableIDMarkerFilePrefix}, func(path string, _ int64) error {
+		restoreCommitTs, snapshotBackupTs, parsed := parseLogRestoreTableIDsMarkerFileName(path)
+		if parsed {
+			if filterOutFn(restoreCommitTs, snapshotBackupTs) {
+				return nil
+			}
+		}
+		filenames = append(filenames, path)
```

Comment:
if not parsed, why still need append this file?

Reply:
I'm worried that some external storage make the filename dirty.